src/services/document_service.py

```diff
@@ -413,10 +413,67 @@ def add(
         if not address.path:  # We're adding something to the dataSource itself
             return self._add_document_to_data_source(address.data_source, document, update_uncontained)
 
-        target: Node = self.get_document(address, depth=99)
+        try:
+            target: Node = self.get_document(address, depth=99)
+        except NotFoundException:
+            target = None
 
         if not target:
-            raise NotFoundException(f"Could not find '{address}' in data source '{address.data_source}'")
+            # If target does not exist, there are 2 cases to consider:
+            #   1) the target is an optional attribute that does not exist yet. In that case, we set the target to be
+            #      the parent of entity referenced by 'address'.
+            #   2) the address is wrong. In that case, raise Exception.
+
+            split_address_path: list[str] = address.path.rsplit(".", 1)
+
+            if len(split_address_path) <= 1 and split_address_path[0] == address.path:
+                # Raising NotFoundException, since the get_document() did not find the document
+                # with address=address.path in the above try except statement
+                raise NotFoundException(f"Could not find document {address}")
+
+            last_attribute_in_address: str = split_address_path.pop()
+            parent_address_as_string: str = ".".join(split_address_path)
```

Comment:
```suggestion
            parent_address_as_string, last_attribute_in_address = split_address
```
I think you don't have to do this pop and join anymore since you did rsplit() and not split() on line 427, so should be enough to do the unpacking since the array will only be two elements long at most.

Reply:
nice!